includes/class-wc-admin-product-category-lookup.php

```diff
@@ -0,0 +1,224 @@
+			return;
+		}
+		$category_ids = array();
+		$term_ids     = wp_get_object_terms( $product_id, 'product_cat', array( 'fields' => 'ids' ) );
+
+		foreach ( $term_ids as $term_id ) {
+			$category_ids[] = $term_id;
+			$category_ids   = array_merge( $category_ids, get_ancestors( $term_id, 'product_cat', 'taxonomy' ) );
+		}
+
+		$category_ids = array_map( 'intval', array_unique( array_filter( $category_ids ) ) );
+		$inserts      = array();
+
+		foreach ( $category_ids as $category_id ) {
+			$inserts[] = $wpdb->prepare( '(%d,%d)', $product_id, $category_id );
+		}
+
+		$insert_query = implode( ',', $inserts );
+
+		$wpdb->query( "INSERT IGNORE INTO {$wpdb->prefix}wc_product_category_lookup (product_id,category_id) VALUES {$insert_query}" );  // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.NotPrepared
+	}
+
+	/**
+	 * Store edits to categories.
+	 *
+	 * @param int $category_id Term ID being edited.
+	 */
+	public function edit_product_cat( $category_id ) {
+		$category = get_term( $category_id, 'product_cat' );
+
+		// Store the parent before the category is updated.
+		$this->edited_product_cats[ $category_id ] = $category->parent;
+	}
+
+	/**
+	 * Modify relationships for a product cat when it's parent is modified.
+	 *
+	 * @param int $category_id Term ID being edited.
+	 */
+	public function edited_product_cat( $category_id ) {
+		global $wpdb;
+
+		if ( ! isset( $this->edited_product_cats[ $category_id ] ) ) {
+			return;
+		}
+
+		$category    = get_term( $category_id, 'product_cat' );
+		$prev_parent = $this->edited_product_cats[ $category_id ];
+		$new_parent  = $category->parent;
+
+		// No edits - no need to modify relationships.
+		if ( $prev_parent === $new_parent ) {
+			unset( $this->edited_product_cats[ $category_id ] );
+			return;
+		}
+
+		// Delete relationships for previous parent and current parent ancestors.
+		$term_ids = array( $prev_parent, $new_parent );
+
+		if ( $prev_parent > 0 ) {
+			$term_ids = $term_ids + get_ancestors( $prev_parent, 'product_cat', 'taxonomy' );
+		}
+
+		if ( $new_parent > 0 ) {
+			$term_ids = $term_ids + get_ancestors( $new_parent, 'product_cat', 'taxonomy' );
+		}
+
+		$term_ids = array_unique( array_filter( $term_ids ) );
+
+		$this->delete_relationships( $term_ids );
+
+		// Now the hierarchy was deleted, we need to re-insert relationships for the previous parent and current parent ancestors and descendents.
+		$children   = _get_term_hierarchy( 'product_cat' );
+		$insert_ids = $term_ids;
+
+		foreach ( $term_ids as $term_id ) {
+			if ( isset( $children[ $term_id ] ) ) {
+				$insert_ids = array_merge( $insert_ids, $children[ $term_id ] );
+			}
+		}
+		$this->move_term_relationships( $insert_ids );
+
+		unset( $this->edited_product_cats[ $category_id ] );
+	}
+
+	/**
+	 * Sync product on save.
+	 *
+	 * @param  int    $post_id Post ID being saved.
+	 * @param  object $post Post object.
+	 */
+	public function sync_products( $post_id, $post ) {
```

Comment:
Any performance concerns for running this on the save events vs using an action scheduler job?

Reply:
@timmyc Since I just wanted to get this up and running I didn't follow existing patterns so I'm open to modifying approach. The only thing I can think of is, as a lookup table it's probably better to be current, and by not deferring to AS you can run it on a per-product basis only when changes occur.

Reply:
(if this moved to core it would be in the product data store and run if categories are changed)

Reply:
> as a lookup table it's probably better to be current

Great point, I think queuing reporting-only lookup table operations is a bit "safer" but any lookup table that can impact other admin/front-end views it might be wise to just take the extra queries right away.

Reply:
Most stores won't be changing categories that often. With this, the category hierarchy will already be in memory (for those without persistent caching) so performance wise doing the direct save might be better than queueing to AS.